modules/javafx.web/src/main/native/Source/WebCore/bindings/java/EventListenerManager.cpp

```diff
@@ -0,0 +1,116 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "EventListenerManager.h"
+#include "JavaEventListener.h"
+#include "DOMWindow.h"
+
+namespace WebCore {
+
+EventListenerManager& EventListenerManager::get_instance()
+{
+    static NeverDestroyed<EventListenerManager> sharedManager;
+    return sharedManager;
+}
+
+void EventListenerManager::registerListener(JavaEventListener *ptr, const JLObject &listener)
+{
+    JavaObjectWrapperHandler *temp_ref = new JavaObjectWrapperHandler(listener);
+    std::pair<JavaEventListener*, JavaObjectWrapperHandler*> entry{ ptr, temp_ref };
+    listener_lists.insert(entry);
+}
+
+void EventListenerManager::unregisterListener(JavaEventListener *ptr)
+{
+     std::map<JavaEventListener*, JavaObjectWrapperHandler*>::iterator it;
+     it = listener_lists.find(ptr);
+
+     if (it != listener_lists.end()) {
+         if (it->second && it->second->use_count() == 1) {
+             delete it->second;
+             it->second = nullptr;
+             listener_lists.erase(it); // remove from list
+         }
+
+         if (it->second && it->second->use_count() > 1)
+             it->second->dref();
+     }
+}
+
+JGObject EventListenerManager::get_listener(JavaEventListener *ptr)
+{
+    std::map<JavaEventListener*, JavaObjectWrapperHandler*>::iterator it;
+    it = listener_lists.find(ptr);
+    if (it != listener_lists.end())
+        return it->second->get_listener();
+
+    return nullptr;
+}
+
+void EventListenerManager::registerDOMWindow(DOMWindow* window, JavaEventListener *ptr)
+{
+    std::map<JavaEventListener*, JavaObjectWrapperHandler*>::iterator it;
+    it = listener_lists.find(ptr);
+    if (it != listener_lists.end())
+        it->second->ref();
+
+    std::pair<JavaEventListener*, DOMWindow*> entry{ ptr, window};
+    windowHasEvent.insert(entry);
+}
+
+void EventListenerManager::unregisterDOMWindow(DOMWindow* window)
+{
+    std::multimap<JavaEventListener*, DOMWindow*>::iterator win_it;
+    for (win_it = windowHasEvent.begin(); win_it != windowHasEvent.end(); win_it++) {
+        // de register associated event listeners with window
+        if (window == win_it->second) {
+            unregisterListener(win_it->first);
+        }
+    }
+}
+
+void EventListenerManager::resetDOMWindow(DOMWindow* window)
+{
+    std::multimap<JavaEventListener*, DOMWindow*>::iterator win_it;
+    std::map<JavaEventListener*, JavaObjectWrapperHandler*>::iterator it;
+    bool isReferringToOtherListener = true;
+
+    for (win_it = windowHasEvent.begin(); win_it != windowHasEvent.end(); win_it++) {
+        it = listener_lists.find(win_it->first);
+        if (window == win_it->second && it->second && it != listener_lists.end() && it->second->use_count() == 1)
+            isReferringToOtherListener = false;
+        else
+            isReferringToOtherListener = true;
+    }
```

Comment:
I think I now understand what this is trying to do, but it doesn't looks like it's implemented correctly nor is it optimal.

It seems that the `isReferringToOtherListener` flag is intended to be true iff there is any `JavaEventListener` with a ref count > 1 associated with the Window being unregistered. Ignoring any possible bugs in how it is implemented, there are two problems with this approach. First, you want to remove entries associated with a particular listener if _that_ listener isn't used by another window. So a global "is any listener referring to another window" isn't what you want. Second, since this is multimap, it would seem better to remove individual `(key,value)` pairs associated with the specific window being removed rather than calling erase only for those listeners that aren't being shared at all. If this is feasible, then you wouldn't have to even care if that listener is used by a node in another window. I'm not familiar enough with C++ multimap calls to know how easy this would be, but presumably, it shouldn't be too hard.

Not directly related to this, it might be cleaner to move this logic to `unregisterDOMWindow` instead of having a separate `resetDOMWindow` method, since this is really part of the same conceptual operation.

Reply:
I have used the suggested method , and tested it works expected. Thanks